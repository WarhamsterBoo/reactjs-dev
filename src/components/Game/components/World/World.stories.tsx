import React from "react";
import { World, CreatureState } from "./World";
import { withKnobs, object } from "@storybook/addon-knobs";

export default { title: "World component", decorators: [withKnobs] };

export const WorldStory: React.FC<{}> = () => {
  const creatures: CreatureState[][] = object("Creatures", [
    [{ IsAlive: true }, { IsAlive: true }, { IsAlive: true }],
    [{ IsAlive: true }, { IsAlive: false }, { IsAlive: false }],
    [{ IsAlive: true }, { IsAlive: true }, { IsAlive: false }],
  ]);
  return <World creatures={creatures} />;
};
