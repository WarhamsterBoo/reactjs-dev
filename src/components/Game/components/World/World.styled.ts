import styled from "@emotion/styled";

export const WorldWrapper = styled.div`
  display: inline-block;
  padding: 1px;
  border: 2px solid lightgray;
`;
