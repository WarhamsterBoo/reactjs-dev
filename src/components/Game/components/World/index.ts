export * from "./World";
export * from "./components";
