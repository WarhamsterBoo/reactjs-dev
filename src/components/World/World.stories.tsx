import { withKnobs } from "@storybook/addon-knobs";
import React from "react";
import { World } from "./World";

export default { title: "Hello component", decorators: [withKnobs] };

export const HelloStory: React.FC<{}> = () => {
  return <World />;
};
