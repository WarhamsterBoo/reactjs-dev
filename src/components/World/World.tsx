import React from "react";

export const World = () => <h1>Hello!</h1>;
