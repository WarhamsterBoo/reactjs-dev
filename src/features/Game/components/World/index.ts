export * from "./components";
export * from "./World";
