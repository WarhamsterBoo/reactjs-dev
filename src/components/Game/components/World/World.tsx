import React from "react";
import { Creature } from "./components";
import { WorldWrapper } from "./World.styled";

export interface CreatureState {
  xCoordinate: number;
  yCoordinate: number;
  IsAlive: boolean;
}

export interface WorldProps {
  creatures: CreatureState[][];
  onClick: () => void;
}

export const World: React.FC<WorldProps> = ({ creatures, onClick }) => {
  if (creatures.length == 0 || creatures[0].length == 0) {
    return null;
  }

  return (
    <WorldWrapper>
      {creatures.map((row, y) => [
        ...row.map((creature, x) => (
          <Creature
            key={`{${x}${y}}`}
            x={creature.xCoordinate}
            y={creature.yCoordinate}
            IsAlive={creature.IsAlive}
            onClick={onClick}
          />
        )),
        <br key={`${y}`} />,
      ])}
    </WorldWrapper>
  );
};
