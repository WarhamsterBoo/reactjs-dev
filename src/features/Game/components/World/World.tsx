import React from "react";
import { Creature } from "./components";
import { WorldWrapper } from "./World.styled";

export interface CreatureState {
  isAlive: boolean;
}

export interface WorldProps {
  creatures: CreatureState[][];
  onClick: (x: number, y: number) => void;
}

export const World: React.FC<WorldProps> = ({ creatures, onClick }) => {
  if (creatures.length == 0 || creatures[0].length == 0) {
    return null;
  }

  return (
    <WorldWrapper>
      {creatures.map((row, y) => [
        ...row.map((creature, x) => (
          <Creature
            key={`{${y}${x}}`}
            x={y}
            y={x}
            isAlive={creature.isAlive}
            onClick={onClick}
          />
        )),
        <br key={`${y}`} />,
      ])}
    </WorldWrapper>
  );
};
