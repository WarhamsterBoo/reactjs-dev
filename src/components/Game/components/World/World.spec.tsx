import React from "react";
import renderer from "react-test-renderer";
import { CreatureState, World } from "./World";
import { mount } from "enzyme";

describe("World", () => {
  it.each`
    creatures
    ${[]}
    ${[[]]}
  `("should render nothing with 0 x 0 size: $creatures", ({ creatures }) => {
    const sut = <World creatures={creatures} onClick={jest.fn()} />;

    expect(renderer.create(sut).toJSON()).toBeNull();
  });

  it("should render with 1 x 1 size and creature is dead", () => {
    const creatures: CreatureState[][] = [[{ IsAlive: false }]];

    const sut = <World creatures={creatures} onClick={jest.fn()} />;

    expect(renderer.create(sut).toJSON()).toMatchSnapshot();
  });

  it("should render with 2 x 2 size filled with creatures", () => {
    const creatures: CreatureState[][] = [
      [{ IsAlive: true }, { IsAlive: true }, { IsAlive: true }],
      [{ IsAlive: true }, { IsAlive: false }, { IsAlive: false }],
      [{ IsAlive: true }, { IsAlive: true }, { IsAlive: false }],
    ];

    const sut = <World creatures={creatures} onClick={jest.fn()} />;

    expect(renderer.create(sut).toJSON()).toMatchSnapshot();
  });

  it("should pass onClick callback to cells", () => {
    const creatures: CreatureState[][] = [[{ IsAlive: false }]];
    const fakeOnClick = jest.fn();
    const sut = mount(<World creatures={creatures} onClick={fakeOnClick} />);

    sut.find("button").simulate("click");

    expect(fakeOnClick).toBeCalledTimes(1);
  });
});
