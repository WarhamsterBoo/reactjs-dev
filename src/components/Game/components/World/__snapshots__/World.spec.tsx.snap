// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`World should render with 1 x 1 size and creature is dead 1`] = `
.emotion-1 {
  display: inline-block;
  padding: 1px;
  background: #2d302b;
}

.emotion-0 {
  width: 20px;
  height: 20px;
  border: none;
  margin: 1px;
  display: inline-block;
  line-height: 25px;
  text-align: center;
  vertical-align: bottom;
  background: #454a43;
}

<div
  className="emotion-1"
>
  <button
    className="emotion-0"
    onClick={[MockFunction]}
  />
  <br />
</div>
`;

exports[`World should render with 2 x 2 size filled with creatures 1`] = `
.emotion-9 {
  display: inline-block;
  padding: 1px;
  background: #2d302b;
}

.emotion-4 {
  width: 20px;
  height: 20px;
  border: none;
  margin: 1px;
  display: inline-block;
  line-height: 25px;
  text-align: center;
  vertical-align: bottom;
  background: #454a43;
}

.emotion-0 {
  width: 20px;
  height: 20px;
  border: none;
  margin: 1px;
  display: inline-block;
  line-height: 25px;
  text-align: center;
  vertical-align: bottom;
  background: #41ff00;
  border-radius: 5px;
}

<div
  className="emotion-9"
>
  <button
    className="emotion-0"
    onClick={[MockFunction]}
  />
  <button
    className="emotion-0"
    onClick={[MockFunction]}
  />
  <button
    className="emotion-0"
    onClick={[MockFunction]}
  />
  <br />
  <button
    className="emotion-0"
    onClick={[MockFunction]}
  />
  <button
    className="emotion-4"
    onClick={[MockFunction]}
  />
  <button
    className="emotion-4"
    onClick={[MockFunction]}
  />
  <br />
  <button
    className="emotion-0"
    onClick={[MockFunction]}
  />
  <button
    className="emotion-0"
    onClick={[MockFunction]}
  />
  <button
    className="emotion-4"
    onClick={[MockFunction]}
  />
  <br />
</div>
`;
